recent_conversations: Exclude muted user PMs from recent conversations.

Ujjawal3 opened the pull request:
Earlier muted private conversations were shown in recent conversations. This commit exclude them from recent conversations.

<!-- Describe your pull request here.-->

Fixes: #24299

<!-- If the PR makes UI changes, always include one or more still screenshots to demonstrate your changes. If it seems helpful, add a screen capture of the new functionality as well.

Tooling tips: https://zulip.readthedocs.io/en/latest/tutorials/screenshot-and-gif-software.html
-->

**Screenshots and screen captures:**

<details>
<summary>Self-review checklist</summary>

<!-- Prior to submitting a PR, follow our step-by-step guide to review your own code:
https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code -->

<!-- Once you create the PR, check off all the steps below that you have completed.
If any of these steps are not relevant or you have not completed, leave them unchecked.-->

- [ ] [Self-reviewed](https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code) the changes for clarity and maintainability
      (variable names, code reuse, readability, etc.).

Communicate decisions, questions, and potential concerns.

- [x] Explains differences from previous plans (e.g., issue description).
- [x] Highlights technical choices and bugs encountered.
- [x] Calls out remaining decisions and concerns.
- [x] Automated tests verify logic where appropriate.

Individual commits are ready for review (see [commit discipline](https://zulip.readthedocs.io/en/latest/contributing/commit-discipline.html)).

- [x] Each commit is a coherent idea.
- [x] Commit message(s) explain reasoning and motivation for changes.

Completed manual review and testing of the following:

- [ ] Visual appearance of the changes.
- [ ] Responsiveness and internationalization.
- [ ] Strings and tooltips.
- [ ] End-to-end functionality of buttons, interactions and flows.
- [ ] Corner cases, error conditions, and easily imagined bugs.
</details>

Ujjawal3 commented:
@alya please review this PR.

alya commented:
Thanks! It looks good to me for conversations where all users are muted, including live updates.

I think for group PMs where some of the users are muted, we should show the names of muted users as "Muted user", like we do in the message header bar:

<img width="722" alt="Screen Shot 2023-03-12 at 11 32 43 PM" src="https://user-images.githubusercontent.com/2090066/224625222-7a283327-dbc4-4920-8823-8175a354ab74.png">
<img width="350" alt="Screen Shot 2023-03-12 at 11 32 10 PM" src="https://user-images.githubusercontent.com/2090066/224625224-ca809de3-1fb1-4cd0-886d-ef0299f69891.png">

alya commented:
Glad to see us fixing this bug. :)

Ujjawal3 commented:
> Thanks! It looks good to me for conversations where all users are muted, including live updates.
> 
> I think for group PMs where some of the users are muted, we should show the names of muted users as "Muted user", like we do in the message header bar:
> 
> <img alt="Screen Shot 2023-03-12 at 11 32 43 PM" width="722" src="https://user-images.githubusercontent.com/2090066/224625222-7a283327-dbc4-4920-8823-8175a354ab74.png">
> 
> <img alt="Screen Shot 2023-03-12 at 11 32 10 PM" width="350" src="https://user-images.githubusercontent.com/2090066/224625224-ca809de3-1fb1-4cd0-886d-ef0299f69891.png">

I have updated the PR and addressed the change suggested.

abhijeetbodas2001 commented on web/src/recent_topics_ui.js:
```diff
@@ -415,10 +416,11 @@ function format_conversation(conversation_data) {
             .map((user) =>
                 render_user_with_status_icon({
-                    name: user.full_name,
+                    name: people.get_display_full_name(user.id),
```
Could you split this out into a separate commit? This seems not related to the main bug fix.

abhijeetbodas2001 replied:
Oh! I see this is for showing "Muted User" in huddles.
Even in that case, I think this would be better in it's own commit, with an appropriate commit message :)

abhijeetbodas2001 commented:
@Ujjawal3 thanks for working on this! I left a couple of comments above.

Ujjawal3 commented on web/src/recent_topics_ui.js:
```diff
@@ -552,6 +554,13 @@ export function filters_should_hide_topic(topic_data) {
 
+    if (filters.has("include_private") && topic_data.type === "private") {
+        const recipients = msg.to_user_ids.split(",").map((id) => Number.parseInt(id, 10));
+        if (recipients.every((id) => muted_users.is_user_muted(id))) {
```
I looked into the code base but could not find any function in `util.ts`. There is no file named `util.js`. I looked for function in other files too but could not find it. I found this function in `pm_conversations`


```
    const recipients = conversation.user_ids_string.split(",").map((id) => Number.parseInt(id, 10));

    if (recipients.every((id) => muted_users.is_user_muted(id))) {
        return false;
    }

    return true;
```
But we cannot use this as it  expects `user_ids_string` as object key but we have `to_user_ids`.

abhijeetbodas2001 replied:
Yes, I meant `util.ts`.

abhijeetbodas2001 replied:
I found `web/src/people.js/split_to_ints` which we can probably use here
```bash
code/zulip > rg "\.split\(\",\"\)\.map\(" web/src/
web/src/pm_conversations.js
18:    const recipients = conversation.user_ids_string.split(",").map((id) => Number.parseInt(id, 10));

web/src/people.js
54:    return lst.split(",").map((s) => Number.parseInt(s, 10));
```

abhijeetbodas2001 commented:
We should also add a test for this new logic in `web/tests/recent_topics.test.js`. Thanks!

abhijeetbodas2001 commented on web/src/recent_topics_ui.js:
```diff
@@ -17,6 +17,7 @@ import * as message_store from "./message_store";
 import * as message_view_header from "./message_view_header";
 import * as muted_topics_ui from "./muted_topics_ui";
+import * as muted_users from "./muted_users";
 import * as narrow from "./narrow";
```
Where is this import used?

Ujjawal3 replied:
Yes, it should be in a different commit. I have changed it now. Also I have added test coverage now.

abhijeetbodas2001 commented on web/tests/recent_topics.test.js:
```diff
@@ -152,17 +181,36 @@ mock_esm("../src/unread", {
+mock_esm("../src/muted_users", {
+    is_user_muted(user_id) {
+        if (user_id === 5) {
+            return false;
```
Why not use the actual functions here (from `user_topics.js`) instead of patching these? The muted topic store is pretty simple (without much dependencies to worry about). See `user_topics.test.js` for example tests where the actual `user_topics.js` code is used.

abhijeetbodas2001 reviewed (commented):
Thanks for the update! The main code looks good to me. I glanced at the test as well, and have left a quick comment below. I'll also be taking a closer look at the test in some time (I'm not that familiar with the `recent_topics` module, so will have to spend some time reading the test code first.)
One thing which we should do is to add the test in the same commit which fixes this bug. That way, the commit can be safely merged at once. (You may need to introduce the test in the first commit and expand on it in the second commit, since the 2 commits (for 1:1 PMs and huddles) fix different bugs.)
Thanks!

Ujjawal3 commented:
@abhijeetbodas2001 I have addressed your feedback. Please have a look.

abhijeetbodas2001 reviewed (commented):
Thanks for the update! I left a small comment above, but looks good to me otherwise.

Ujjawal3 commented on web/tests/recent_topics.test.js:
```diff
@@ -436,6 +468,48 @@ test("test_filter_all", ({mock_template}) => {
 
+test("test_filter_pm", ({mock_template}) => {
+    page_params.is_spectator = true;
+    const expected = {
```
Yes, it doesn't matter what value we give to `is_spectator`. So I made it false.

timabbott commented:
The large amount of mocking in the `recent_topics_ui.js` tests is already a problem; posted a suggestion on how to avoid making it worse. But I think longer term, we're going to need to rewrite that test to start with a set of normal message objects and mocking way fewer data modules.

Ujjawal3 commented on web/tests/recent_topics.test.js:
```diff
@@ -88,7 +90,12 @@ mock_esm("../src/message_list_data", {
+            return messages[msg_id - 1];
+        }
+        return private_messages[msg_id - 12];
+    },
```
Putting the new messages in `messages` will make other tests to fail. Because in many tests we are processing all the messages inside `messages`.
```
web/tests/recent_topics.test.js:427:    rt.process_messages(messages);
web/tests/recent_topics.test.js:587:    rt.process_messages(messages);
web/tests/recent_topics.test.js:708:    rt.process_messages(messages);
web/tests/recent_topics.test.js:789:    rt.process_messages(messages);
web/tests/recent_topics.test.js:809:    rt.process_messages(messages);
web/tests/recent_topics.test.js:937:    rt.process_messages(messages);
web/tests/recent_topics.test.js:986:    rt.process_messages(messages);
web/tests/recent_topics.test.js:1028:    rt.process_messages(messages);
```
So , I thought it will be better to create a new variable for private conversations instead of changing existing tests.

timabbott replied:
I think it'd probably be better long-term to add them to `messages`, and just extend the verification in those existing tests to expect additional content.

If it takes more than an hour to do, we can potentially decide to merge this for now and plan that as a follow-up project.

timabbott commented:
OK, merged; I think we do want to rewrite the "recent topics" tests to just use a larger base set of messages that include private messages too, and more generally mock 5x fewer modules (in favor of just initializing them with mock data), but that's a longer-term project that we don't want to block this bug fix. @amanagr FYI, though I think cleaning up this test suite is not worth interrupting your other work.
